File: mobile/locket_ai/lib/models/friendship.dart
import 'package:equatable/equatable.dart';
import 'user.dart';

enum FriendshipStatus { pending, accepted, blocked }

class Friendship extends Equatable {
  final String friendshipId;
  final User? userOne;
  final User? userTwo;
  final FriendshipStatus status;
  final DateTime createdAt;

  const Friendship({
    required this.friendshipId,
    this.userOne,
    this.userTwo,
    required this.status,
    required this.createdAt,
  });

  factory Friendship.fromJson(Map<String, dynamic> json) {
    return Friendship(
      friendshipId: json['friendshipId'] ?? '',
      userOne:
          json['userOne'] != null ? User.fromJson(json['userOne']) : null,
      userTwo:
          json['userTwo'] != null ? User.fromJson(json['userTwo']) : null,
      status: _statusFromString(json['status']),
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'friendshipId': friendshipId,
      'userOne': userOne?.toJson(),
      'userTwo': userTwo?.toJson(),
      'status': status.name.toUpperCase(),
      'createdAt': createdAt.toIso8601String(),
    };
  }

  Friendship copyWith({
    String? friendshipId,
    User? userOne,
    User? userTwo,
    FriendshipStatus? status,
    DateTime? createdAt,
  }) {
    return Friendship(
      friendshipId: friendshipId ?? this.friendshipId,
      userOne: userOne ?? this.userOne,
      userTwo: userTwo ?? this.userTwo,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  static FriendshipStatus _statusFromString(String? value) {
    switch (value?.toUpperCase()) {
      case 'ACCEPTED':
        return FriendshipStatus.accepted;
      case 'BLOCKED':
        return FriendshipStatus.blocked;
      default:
        return FriendshipStatus.pending;
    }
  }

  @override
  List<Object?> get props => [friendshipId, userOne, userTwo, status, createdAt];
}

File: backend/src/main/java/com/pbl6/backend/model/Friendship.java
package com.pbl6.backend.model;

import jakarta.persistence.*;
import org.hibernate.annotations.GenericGenerator;
import java.time.LocalDateTime;

@Entity
@Table(name = "Friendships")
public class Friendship {
    
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "friendship_id", length = 36)
    private String friendshipId;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_one_id", nullable = false)
    private User userOne;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_two_id", nullable = false)
    private User userTwo;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private FriendshipStatus status = FriendshipStatus.PENDING;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "action_user_id")
    private User actionUser;
    
    public enum FriendshipStatus {
        PENDING, ACCEPTED, BLOCKED
    }
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
    
    // Constructors
    public Friendship() {}
    
    public Friendship(User userOne, User userTwo) {
        this.userOne = userOne;
        this.userTwo = userTwo;
    }
    
    public Friendship(User userOne, User userTwo, FriendshipStatus status) {
        this.userOne = userOne;
        this.userTwo = userTwo;
        this.status = status;
    }
    
    // Getters and Setters
    public String getFriendshipId() {
        return friendshipId;
    }
    
    public void setFriendshipId(String friendshipId) {
        this.friendshipId = friendshipId;
    }
    
    public User getUserOne() {
        return userOne;
    }
    
    public void setUserOne(User userOne) {
        this.userOne = userOne;
    }
    
    public User getUserTwo() {
        return userTwo;
    }
    
    public void setUserTwo(User userTwo) {
        this.userTwo = userTwo;
    }
    
    public FriendshipStatus getStatus() {
        return status;
    }
    
    public void setStatus(FriendshipStatus status) {
        this.status = status;
    }
    
    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
    
    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public User getActionUser() {
        return actionUser;
    }

    public void setActionUser(User actionUser) {
        this.actionUser = actionUser;
    }
}